### DISCUSSION TITLE
GravNet example

### DISCUSSION THREAD

**Discussion by tferber** [pyg-team/pytorch_geometric#2214] [Q&A] [answered]
I have implemented a three-target variant of https://arxiv.org/pdf/1902.07987.pdf in pytorch.geometric: https://github.com/tferber/ECLML/blob/main/models/models.py using the GravNetConv layer (https://pytorch-geometric.readthedocs.io/en/latest/_modules/torch_geometric/nn/conv/gravnet_conv.html) that should simultaneously detect target classes and background fluctuations.

With any reasonable learning rates and model parameters the model will detect class 3 (gray), but more or less assign 50/50 weights to classes 1 and 2 (blue and red): https://desycloud.desy.de/index.php/s/9cGfdECbaP2dx7g

With a single event i can get the model to overtrain (https://desycloud.desy.de/index.php/s/w49B2mWbDEyRsRM) which makes me think that it works somewhat..., but as soon as i use a few events in a batch, the performance goes does significantly. Maybe I have a problem with the sparse block diagonal adjacency matrices?

I wonder if there is any working example/test of the pytorch GravNetConv layer?

**Comment by rusty1s** [comment, marked as answer]
Sadly, we do not have a GravNetConv example yet (pinging @jkiesele just in case he wants to provide one :))

In general, block diagonal adjacency matrices should be automatically created using `torch_geometric.data.DataLoader`. Can you show a small example? This might help to identify the issue.

**Reply by jkiesele** [reply]
We might, in fact, provide an example soon (btw, we are also working slowly on porting our kNN and accumulate kernels, I'll keep you posted on that). However, this won't happen in the next days.
The adjacency matrix is created on the fly in GravNet using the same implementation as EdgeConv, so this part should work, in principle.

Looking at the plots, there might be a conceptual issue: how do you define which photon is p0 and which is p1? If there is no clear rule for doing so on truth level, the network won't have a chance to learn it properly (but would on only one event), and would probably revert to predicting 50/50.
So we are entering the multi-object regime here, even if it is a fixed number of two.

**Reply by tferber** [reply]
Moin @rusty1s, this is not minimal, but all code is in https://github.com/tferber/ECLML, specifically the dataset is here https://github.com/tferber/ECLML/blob/main/datasets/datasets.py. I can also provide an input file if that helps debugging. 

I am working on the truth level recommendation by @jkiesele now and will let you know if that solves the problem.

**Reply by tferber** [reply]
Fixing the truth level by ordering the targets by azimuthal angle did the trick, or at least a large part of it: https://desycloud.desy.de/index.php/s/ga33JBCyzst73WX 

Thanks!
